In [1]:
import sys  
!{sys.executable} -m pip install matplotlib python-dateutil

You should consider upgrading via the '/usr/local/Cellar/jupyterlab/2.2.0/libexec/bin/python3.8 -m pip install --upgrade pip' command.


In [2]:
import calendar
import datetime as dt
import dateutil.parser as dp
import pandas as pd
import matplotlib.pyplot as plt
from functools import partial

In [3]:
df = pd.read_json(r'data.json', convert_dates=True)
df["duration"] = (df["end_time"] - df["start_time"]).dt.days
df["count"] = 1
df.head()

q_75 = partial(pd.Series.quantile, q=0.75)
q_75.__name__ = "75%"

In [4]:
cancelled = df[(df["idList"] == "Cancelled")]
done = df[(df["idList"] == "Done")]

# Raw data...

In [5]:
df

name  \
0                                          MCAFEE_EPO   
1                                            BRO_JSON   
2                                           DARKTRACE   
3                                            ABSOLUTE   
4                                       CLOUD_PASSAGE   
5                                      SOURCEFIRE_IDS   
6   Fix Attached Files and security_result to GMAI...   
7                         AIRWATCH (log issues fixed)   
8                                       IPSWITCH_SFTP   
9   Reduce significant alerts generated by GMAIL_LOGS   
10                                            CLAM_AV   
11  GMAIL_LOGS: Fix -> Change "Too many recipients...   
12  MICROSOFT_CASB: Fix -> Use go/udm:cloudy Event...   
13                                         GMAIL_LOGS   
14               GCLOUD_AUDIT new CLF specific parser   
15                                     F5_ASM default   
16                          GSUITE_AUDIT: user.userid   
17                   GCP CLOUD_AUDIT: Fix user.userid   
18                            OPENAM: Fix user.userid   
19                        MICROSOFT_CASB: user.userid   
20             CISCO_WIRELESS (Flaky tests debugging)   
21                                     CISCO_WIRELESS   
22                                                IIS   
23                                         SECURELINK   
24  ADAUDIT: "USER_* event types - Change user inf...   
25                                             OPENAM   
26  CYBERARK: USER_* event types - Move user ID fr...   
27                                               OKTA   
28                                         MCAFEE_IPS   
29  CENTRIFY_SSO: "CBN - Some parsers put email in...   
30                                      MICROSOFT_SQL   
31           PREEMPT: Fix event type to SCAN_NETWORK.   
32                                        JUNIPER_IPS   
33                                            PREEMPT   
34                              VMWARE_ESX [IRM-JSON]   
35                                            ALGOSEC   
36                                             BOMGAR   
37                                       ADAUDIT_PLUS   
38                                         VMWARE_ESX   
39                               WINDOWS_DEFENDER_ATP   
40                                           CYBERARK   
41                                          CISCO_ISE   
42                                           DUO_AUTH   
43                                          CLEARPASS   
44                                       CISCO_SWITCH   
45                                      MIMECAST_MAIL   
46                                     THINKST_CANARY   
47                         CLEARPASS: Fix user.userid   
48                    MICROSOFT_SECURITY_CENTER_ALERT   
49                                           AIRWATCH   
50                                             F5_ASM   
51                    AWS_CLOUDTRAIL [ListAccessKeys]   
52                                     VMWARE_VCENTER   
53                                      EXTRAHOP_DHCP   
54                                           THYCOTIC   
55                                     PROOFPOINT_TAP   
56                                         CLOUDFLARE   

                         shortUrl     idList                start_time  \
0   https://trello.com/c/k60VLjOa       Done 2020-08-24 13:22:00+00:00   
1   https://trello.com/c/Eui55N6L       Done 2020-09-03 19:42:51+00:00   
2   https://trello.com/c/8wx0tVYa       Done 2020-09-01 19:11:55+00:00   
3   https://trello.com/c/FVqlbJVA       Done 2020-08-28 15:17:54+00:00   
4   https://trello.com/c/zvLnKRMt       Done 2020-08-19 16:58:00+00:00   
5   https://trello.com/c/zGAtO5Io       Done 2020-07-30 17:34:08+00:00   
6   https://trello.com/c/aeMaPZDe       Done 2020-08-13 19:33:45+00:00   
7   https://trello.com/c/XpFvP9Xm       Done 2020-07-28 15:33:23+00:00   
8   https://trello.com/c/H76wnXs7       Done 2020-07-17 15:12:18+00:00   
9 

# Done estimate

In [6]:
done["duration"].describe()
done["duration"].agg([q_75])

75%    21.0
Name: duration, dtype: float64

# Trying another one: Done estimate (with "extremes" removed)

In [7]:
done_extremes_removed = done[(done["duration"] > 5) & (done["duration"] < 35)]
#done_extremes_removed.sort_values(by="duration")

In [8]:
done_extremes_removed["duration"].describe()
done_extremes_removed["duration"].agg([q_75])

75%    21.75
Name: duration, dtype: float64

# Trying another one: Now separating small from non-small

In [9]:
small_extremes_removed = done_extremes_removed[(done_extremes_removed["duration"] <= 10)]
not_small_extremes_removed = done_extremes_removed[(done_extremes_removed["duration"] > 10)]

In [10]:
small_extremes_removed["duration"].describe()
small_extremes_removed["duration"].agg([q_75])

75%    8.5
Name: duration, dtype: float64

In [11]:
not_small_extremes_removed["duration"].describe()
not_small_extremes_removed["duration"].agg([q_75])

75%    25.25
Name: duration, dtype: float64

In [12]:
small = done[(done["duration"] <= 6)]
not_small = done[(done["duration"] >= 6)]

# Another try: Don't remove extremes, but separate in small and non-small group

In [13]:
small["duration"].describe()
small["duration"].agg([q_75])

75%    4.75
Name: duration, dtype: float64

In [14]:
not_small["duration"].describe()
not_small["duration"].agg([q_75])

75%    23.5
Name: duration, dtype: float64

# Ok, let's go with 4.75 for small ones, and 23.5 for non-small ones.

## 5 weekdays for small ones

## 17 weekdays (23 days - 6 weekends) for non-small ones